# Домашнее задание № 1

In [156]:
from razdel import sentenize
from razdel import tokenize as razdel_tokenize
from string import punctuation as p
from pymystem3 import Mystem
from json import loads
from pymorphy2.analyzer import MorphAnalyzer
from spacy import load
from nltk.tokenize import sent_tokenize
import hashlib
from collections import defaultdict
from tqdm.auto import tqdm

## Задание 1 (2 балла)

Найдите 1 любой способ сломать токенизацию на предложения функцией sentenize из библиотеки razdel. Придумайте (или найдите на каком-то корпусе) такое предложение (или несколько предложений), которое будет некорректно разобрано sentenize, но при этом будет грамматически корректным. 

In [132]:
text = '— Как можно быть здоровой... когда нравственно страдаешь? Разве можно, имея чувство, оставаться спокойною в наше время? — сказала Анна Павловна. — Вы весь вечер у меня, надеюсь?'
sents = list(sentenize(text))
sents

[Substring(0, 57, '— Как можно быть здоровой... когда нравственно страдаешь?'),
 Substring(58,
           144,
           'Разве можно, имея чувство, оставаться спокойною в наше время? — сказала Анна Павловна.'),
 Substring(145, 177, '— Вы весь вечер у меня, надеюсь?')]

(Л.Н. Толстой.) Если считать все законченные знаком вопроса части прямой речи предложениями, то "сказала Анна Павловна" не отделено от предыдущего. Если это всё одно большое -- понятно, что выделить три неверно. Если здесь всего два предложения ("— Вы весь вечер у меня, надеюсь?" -- второе), то три тоже неверно.

## 2. Токенизация Mystem vs razdel.tokenize (2 балла)


Токенизируйте текст с помощью razdel и с помощью Mystem. Найдите различия в токенизациях. Что по вашему работает лучше на приведенном тексте?

In [3]:
text = """
Вторым и третьим открытыми белыми карликами стали Сириус B и Процион B. В 1844 году директор Кёнигсбергской обсерватории Фридрих Бессель, анализируя данные наблюдений, которые велись с 1755 года, обнаружил, что Сириус, ярчайшая звезда земного неба, и Процион периодически, хотя и весьма слабо, отклоняются от прямолинейной траектории движения по небесной сфере[5]. Бессель пришёл к выводу, что у каждой из них должен быть близкий спутник. Сообщение было встречено скептически, поскольку слабый спутник оставался ненаблюдаемым, а его масса должна была быть достаточно велика — сравнимой с массой Сириуса и Проциона, соответственно.

В январе 1862 года Элвин Грэхэм Кларк, юстируя 18-дюймовый рефрактор, самый большой на то время телескоп в мире (Dearborn Telescope), впоследствии поставленный семейной фирмой Кларков в обсерваторию Чикагского университета, обнаружил в непосредственной близости от Сириуса тусклую звёздочку. Это был спутник Сириуса, Сириус B, предсказанный Бесселем[6]. А в 1896 году американский астроном Д. М. Шеберле открыл Процион B, подтвердив тем самым и второе предсказание Бесселя.

В 1915 году американский астроном Уолтер Сидней Адамс измерил спектр Сириуса B. Из измерений следовало, что его температура не ниже, чем у Сириуса A (по современным данным, температура поверхности Сириуса B составляет 25 000 K, а Сириуса A — 10 000 К), что, с учётом его в 10 000 раз меньшей, чем у Сириуса A, светимости указывает на очень малый радиус и, соответственно, высокую плотность — 106 г/см3 (плотность Сириуса ~0,25 г/см3, плотность Солнца ~1,4 г/см3).
"""

Сравню без пунктуации

In [44]:
razdel_tokens = [token.text for token in list(razdel_tokenize(text)) \
                 if token.text not in p + '—']
razdel_tokens

['Вторым',
 'и',
 'третьим',
 'открытыми',
 'белыми',
 'карликами',
 'стали',
 'Сириус',
 'B',
 'и',
 'Процион',
 'B',
 'В',
 '1844',
 'году',
 'директор',
 'Кёнигсбергской',
 'обсерватории',
 'Фридрих',
 'Бессель',
 'анализируя',
 'данные',
 'наблюдений',
 'которые',
 'велись',
 'с',
 '1755',
 'года',
 'обнаружил',
 'что',
 'Сириус',
 'ярчайшая',
 'звезда',
 'земного',
 'неба',
 'и',
 'Процион',
 'периодически',
 'хотя',
 'и',
 'весьма',
 'слабо',
 'отклоняются',
 'от',
 'прямолинейной',
 'траектории',
 'движения',
 'по',
 'небесной',
 'сфере',
 '5',
 'Бессель',
 'пришёл',
 'к',
 'выводу',
 'что',
 'у',
 'каждой',
 'из',
 'них',
 'должен',
 'быть',
 'близкий',
 'спутник',
 'Сообщение',
 'было',
 'встречено',
 'скептически',
 'поскольку',
 'слабый',
 'спутник',
 'оставался',
 'ненаблюдаемым',
 'а',
 'его',
 'масса',
 'должна',
 'была',
 'быть',
 'достаточно',
 'велика',
 'сравнимой',
 'с',
 'массой',
 'Сириуса',
 'и',
 'Проциона',
 'соответственно',
 'В',
 'январе',
 '1862',
 'года',
 

In [52]:
with open('text.txt', 'w', encoding='utf-8') as f:
    f.write(text)

# флаг -c -- сохраняем междусловное
!mystem -c --format json text.txt text_parsed.txt

with open('text_parsed.txt', encoding='utf-8') as f:
    processed_text = [loads(line) for line in f]

mystem_console_tokens = []
for i in processed_text:
    if i == []:
        pass
    else:
        for j in i:
            # '),' -- артефакт подхода
            if j['text'] != ' ' and j['text'] != '\r\n' \
            and j['text'].strip() not in p + '—' + '),':
                mystem_console_tokens.append(j['text'])
mystem_console_tokens

['Вторым',
 'и',
 'третьим',
 'открытыми',
 'белыми',
 'карликами',
 'стали',
 'Сириус',
 'B',
 'и',
 'Процион',
 'B',
 'В',
 '1844',
 'году',
 'директор',
 'Кёнигсбергской',
 'обсерватории',
 'Фридрих',
 'Бессель',
 'анализируя',
 'данные',
 'наблюдений',
 'которые',
 'велись',
 'с',
 '1755',
 'года',
 'обнаружил',
 'что',
 'Сириус',
 'ярчайшая',
 'звезда',
 'земного',
 'неба',
 'и',
 'Процион',
 'периодически',
 'хотя',
 'и',
 'весьма',
 'слабо',
 'отклоняются',
 'от',
 'прямолинейной',
 'траектории',
 'движения',
 'по',
 'небесной',
 'сфере',
 '5',
 'Бессель',
 'пришёл',
 'к',
 'выводу',
 'что',
 'у',
 'каждой',
 'из',
 'них',
 'должен',
 'быть',
 'близкий',
 'спутник',
 'Сообщение',
 'было',
 'встречено',
 'скептически',
 'поскольку',
 'слабый',
 'спутник',
 'оставался',
 'ненаблюдаемым',
 'а',
 'его',
 'масса',
 'должна',
 'была',
 'быть',
 'достаточно',
 'велика',
 'сравнимой',
 'с',
 'массой',
 'Сириуса',
 'и',
 'Проциона',
 'соответственно',
 'В',
 'январе',
 '1862',
 'года',
 

In [53]:
print(len(razdel_tokens))
print(len(mystem_console_tokens))

233
233


In [58]:
for i in range(len(razdel_tokens)):
    if razdel_tokens[i] != mystem_console_tokens[i]:
        print(i, razdel_tokens[i], mystem_console_tokens[i])
        break
for j in range(i + 1, len(razdel_tokens)):
    if razdel_tokens[j] != mystem_console_tokens[j + 1]:
        print(j, razdel_tokens[j], mystem_console_tokens[j + 1])
        break

96 18-дюймовый 18
219 см см3


Видно, что библиотеки по-разному (зеркально) обходятся со случаями, когда между пробелами буквочисловой токен

Смотрю пунктуацию, формально -- not .isalpha()

In [65]:
razdel_punct = [token.text for token in list(razdel_tokenize(text)) \
                if not token.text.isalpha()]
razdel_punct

['.',
 '1844',
 ',',
 ',',
 '1755',
 ',',
 ',',
 ',',
 ',',
 ',',
 ',',
 '[',
 '5',
 ']',
 '.',
 ',',
 '.',
 ',',
 ',',
 '—',
 ',',
 '.',
 '1862',
 ',',
 '18-дюймовый',
 ',',
 '(',
 ')',
 ',',
 ',',
 '.',
 ',',
 ',',
 '[',
 '6',
 ']',
 '.',
 '1896',
 '.',
 '.',
 ',',
 '.',
 '1915',
 '.',
 ',',
 ',',
 '(',
 ',',
 '25',
 '000',
 ',',
 '—',
 '10',
 '000',
 ')',
 ',',
 ',',
 '10',
 '000',
 ',',
 ',',
 ',',
 ',',
 '—',
 '106',
 '/',
 '3',
 '(',
 '~',
 '0,25',
 '/',
 '3',
 ',',
 '~',
 '1,4',
 '/',
 '3',
 ')',
 '.']

In [66]:
with open('text_parsed.txt', encoding='utf-8') as f:
    processed_text = [loads(line) for line in f]

mystem_console_punct = []
for i in processed_text:
    if i == []:
        pass
    else:
        for j in i:
            if j['text'] != ' ' and j['text'] != '\r\n' \
            and not j['text'].isalpha():
                mystem_console_punct.append(j['text'])
mystem_console_punct

['. ',
 '1844',
 ', ',
 ', ',
 '1755',
 ', ',
 ', ',
 ', ',
 ', ',
 ', ',
 ', ',
 '[',
 '5',
 ']',
 '. ',
 ', ',
 '. ',
 ', ',
 ', ',
 ' — ',
 ', ',
 '.',
 '1862',
 ', ',
 '18',
 '-',
 ', ',
 ' (',
 '), ',
 ', ',
 '. ',
 ', ',
 ', ',
 '[',
 '6',
 ']',
 '. ',
 '1896',
 '. ',
 '. ',
 ', ',
 '.',
 '1915',
 '. ',
 ', ',
 ', ',
 ' (',
 ', ',
 '25',
 '000',
 ', ',
 ' — ',
 '10',
 '000',
 '), ',
 ', ',
 '10',
 '000',
 ', ',
 ', ',
 ', ',
 ', ',
 ' — ',
 '106',
 '/',
 'см3',
 ' (',
 ' ~',
 '0',
 ',',
 '25',
 '/',
 'см3',
 ', ',
 ' ~',
 '1',
 ',',
 '4',
 '/',
 'см3',
 ')',
 '.']

Можно видеть, что вообще mystem делает токены из знака препинания и пробела, но сделаем .strip()

In [77]:
for i in range(len(mystem_console_punct)):
    mystem_console_punct[i] = mystem_console_punct[i].strip()

In [71]:
print(len(razdel_punct))
print(len(mystem_console_punct))

79
82


In [79]:
for i in range(len(razdel_punct)):
    if razdel_punct[i] != mystem_console_punct[i]:
        print(i, [razdel_punct[i], mystem_console_punct[i]])
        break
for j in range(i + 1, len(razdel_punct)):
    if razdel_punct[j] != mystem_console_punct[j + 1]:
        print(j, razdel_punct[j], mystem_console_punct[j + 1])
        break

24 ['18-дюймовый', '18']
27 ) ),


Чего-то нового не находим. В целом парсеры работают очень похоже, но отличия всё же есть

## 3. Лемматизация Mystem vs Pymorphy (2 балла)

Лемматизируйте текст с помощью mystem и pymorphy. Найдите различия в лемматизации. Что по вашему работает лучше на приведенном тексте?

Важно: для пайморфи используйте токенизацию из mystem, чтобы исключить влияние токенизации на результат. Анализируйте только значимые различия, а не технические особенности (не сравнивайте скорость работы и удобность интерфейса).

В майстеме убедитесь, что используется дизамбигуация.

In [86]:
mystem = Mystem(disambiguation=True)
mystem_lemmas = mystem.lemmatize(text)
mystem_lemmas

['\n',
 'второй',
 ' ',
 'и',
 ' ',
 'третий',
 ' ',
 'открытый',
 ' ',
 'белый',
 ' ',
 'карлик',
 ' ',
 'становиться',
 ' ',
 'сириус',
 ' ',
 'B',
 ' ',
 'и',
 ' ',
 'процион',
 ' ',
 'B',
 '. ',
 'в',
 ' ',
 '1844',
 ' ',
 'год',
 ' ',
 'директор',
 ' ',
 'кенигсбергский',
 ' ',
 'обсерватория',
 ' ',
 'фридрих',
 ' ',
 'бессель',
 ', ',
 'анализировать',
 ' ',
 'данные',
 ' ',
 'наблюдение',
 ', ',
 'который',
 ' ',
 'вестись',
 ' ',
 'с',
 ' ',
 '1755',
 ' ',
 'год',
 ', ',
 'обнаруживать',
 ', ',
 'что',
 ' ',
 'сириус',
 ', ',
 'яркий',
 ' ',
 'звезда',
 ' ',
 'земной',
 ' ',
 'небо',
 ', ',
 'и',
 ' ',
 'процион',
 ' ',
 'периодически',
 ', ',
 'хотя',
 ' ',
 'и',
 ' ',
 'весьма',
 ' ',
 'слабо',
 ', ',
 'отклоняться',
 ' ',
 'от',
 ' ',
 'прямолинейный',
 ' ',
 'траектория',
 ' ',
 'движение',
 ' ',
 'по',
 ' ',
 'небесный',
 ' ',
 'сфера',
 '[',
 '5',
 ']',
 '. ',
 'бессель',
 ' ',
 'приходить',
 ' ',
 'к',
 ' ',
 'вывод',
 ', ',
 'что',
 ' ',
 'у',
 ' ',
 'каждый',
 ' ',
 '

In [87]:
with open('text_parsed.txt', encoding='utf-8') as f:
    processed_text = [loads(line) for line in f]

mystem_tokenization = []
for i in processed_text:
    if i == []:
        pass
    else:
        for j in i:
            mystem_tokenization.append(j['text'])
mystem_tokenization

['\r\n',
 'Вторым',
 ' ',
 'и',
 ' ',
 'третьим',
 ' ',
 'открытыми',
 ' ',
 'белыми',
 ' ',
 'карликами',
 ' ',
 'стали',
 ' ',
 'Сириус',
 ' ',
 'B',
 ' ',
 'и',
 ' ',
 'Процион',
 ' ',
 'B',
 '. ',
 'В',
 ' ',
 '1844',
 ' ',
 'году',
 ' ',
 'директор',
 ' ',
 'Кёнигсбергской',
 ' ',
 'обсерватории',
 ' ',
 'Фридрих',
 ' ',
 'Бессель',
 ', ',
 'анализируя',
 ' ',
 'данные',
 ' ',
 'наблюдений',
 ', ',
 'которые',
 ' ',
 'велись',
 ' ',
 'с',
 ' ',
 '1755',
 ' ',
 'года',
 ', ',
 'обнаружил',
 ', ',
 'что',
 ' ',
 'Сириус',
 ', ',
 'ярчайшая',
 ' ',
 'звезда',
 ' ',
 'земного',
 ' ',
 'неба',
 ', ',
 'и',
 ' ',
 'Процион',
 ' ',
 'периодически',
 ', ',
 'хотя',
 ' ',
 'и',
 ' ',
 'весьма',
 ' ',
 'слабо',
 ', ',
 'отклоняются',
 ' ',
 'от',
 ' ',
 'прямолинейной',
 ' ',
 'траектории',
 ' ',
 'движения',
 ' ',
 'по',
 ' ',
 'небесной',
 ' ',
 'сфере',
 '[',
 '5',
 ']',
 '. ',
 'Бессель',
 ' ',
 'пришёл',
 ' ',
 'к',
 ' ',
 'выводу',
 ', ',
 'что',
 ' ',
 'у',
 ' ',
 'каждой',
 ' ',
 'и

In [99]:
morph = MorphAnalyzer()

pymorphy_lemmas = []
for i in mystem_tokenization:
    pymorphy_lemmas.append(morph.parse(i)[0].normal_form)
pymorphy_lemmas

['\r\n',
 'второй',
 ' ',
 'и',
 ' ',
 'третий',
 ' ',
 'открытый',
 ' ',
 'белый',
 ' ',
 'карлик',
 ' ',
 'стать',
 ' ',
 'сириус',
 ' ',
 'b',
 ' ',
 'и',
 ' ',
 'процион',
 ' ',
 'b',
 '. ',
 'в',
 ' ',
 '1844',
 ' ',
 'год',
 ' ',
 'директор',
 ' ',
 'кёнигсбергский',
 ' ',
 'обсерватория',
 ' ',
 'фридрих',
 ' ',
 'бессель',
 ', ',
 'анализировать',
 ' ',
 'дать',
 ' ',
 'наблюдение',
 ', ',
 'который',
 ' ',
 'вестись',
 ' ',
 'с',
 ' ',
 '1755',
 ' ',
 'год',
 ', ',
 'обнаружить',
 ', ',
 'что',
 ' ',
 'сириус',
 ', ',
 'яркий',
 ' ',
 'звезда',
 ' ',
 'земной',
 ' ',
 'небо',
 ', ',
 'и',
 ' ',
 'процион',
 ' ',
 'периодически',
 ', ',
 'хотя',
 ' ',
 'и',
 ' ',
 'весьма',
 ' ',
 'слабо',
 ', ',
 'отклоняться',
 ' ',
 'от',
 ' ',
 'прямолинейный',
 ' ',
 'траектория',
 ' ',
 'движение',
 ' ',
 'по',
 ' ',
 'небесный',
 ' ',
 'сфера',
 '[',
 '5',
 ']',
 '. ',
 'бессель',
 ' ',
 'прийти',
 ' ',
 'к',
 ' ',
 'вывод',
 ', ',
 'что',
 ' ',
 'у',
 ' ',
 'каждый',
 ' ',
 'из',
 ' ',


In [106]:
print(len(mystem_lemmas))
print(len(pymorphy_lemmas))

475
475


In [110]:
for i in range(len(mystem_lemmas)):
    if mystem_lemmas[i] != pymorphy_lemmas[i]:
        print(i, [mystem_lemmas[i], pymorphy_lemmas[i]])

0 ['\n', '\r\n']
13 ['становиться', 'стать']
17 ['B', 'b']
23 ['B', 'b']
33 ['кенигсбергский', 'кёнигсбергский']
43 ['данные', 'дать']
57 ['обнаруживать', 'обнаружить']
106 ['приходить', 'прийти']
134 ['встречать', 'встретить']
150 ['его', 'он']
178 ['\n', '\r\n']
179 ['\n', '\r\n']
190 ['грэхэм', 'грэхэма']
218 ['Dearborn', 'dearborn']
220 ['Telescope', 'telescope']
224 ['поставлять', 'поставить']
240 ['обнаруживать', 'обнаружить']
254 ['звездочка', 'звёздочка']
266 ['B', 'b']
268 ['предсказывать', 'предсказать']
287 ['д', 'далее']
291 ['шеберль', 'шеберл']
293 ['открывать', 'открыть']
297 ['B', 'b']
301 ['то', 'тем']
307 ['второй', 'второе']
313 ['\n', '\r\n']
314 ['\n', '\r\n']
331 ['измерять', 'измерить']
337 ['B', 'b']
347 ['его', 'он']
353 ['низкий', 'ниже']
361 ['A', 'a']
375 ['B', 'b']
383 ['K', 'k']
389 ['A', 'a']
401 ['учет', 'учёт']
413 ['меньший', 'малый']
421 ['A', 'a']
474 ['\n', '\r\n']


В целом лучше, как кажется, работает mystem (см. данные, грэхэм), но разница очень незначительная, в основном приходится на спорные случаи или на сложные типа иностранных имён

## 4. Лемматизация в SpaCy (2 балла)

С помощью Spacy (модель для русского языка) лемматизируйте тот же текст. Проверьте есть ли различия с Mystem и Pymoprhy.

In [105]:
nlp = load("ru_core_news_lg")

doc = nlp(text)

spacy_lemmas = []
for sent in doc.sents:
    for token in sent:
        spacy_lemmas.append(token.lemma_)
spacy_lemmas

['\n',
 'второй',
 'и',
 'третий',
 'открытый',
 'белый',
 'карлик',
 'стать',
 'сириус',
 'b',
 'и',
 'процион',
 'b.',
 'в',
 '1844',
 'год',
 'директор',
 'кёнигсбергский',
 'обсерватория',
 'фридрих',
 'бессель',
 ',',
 'анализировать',
 'данные',
 'наблюдение',
 ',',
 'которые',
 'вестись',
 'с',
 '1755',
 'год',
 ',',
 'обнаружить',
 ',',
 'что',
 'сириус',
 ',',
 'ярчайшая',
 'звезда',
 'земной',
 'небо',
 ',',
 'и',
 'процион',
 'периодически',
 ',',
 'хотя',
 'и',
 'весьма',
 'слабо',
 ',',
 'отклоняться',
 'от',
 'прямолинейный',
 'траектория',
 'движение',
 'по',
 'небесный',
 'сфере[5',
 ']',
 '.',
 'бессель',
 'прийти',
 'к',
 'вывод',
 ',',
 'что',
 'у',
 'каждый',
 'из',
 'них',
 'должный',
 'быть',
 'близкий',
 'спутник',
 '.',
 'сообщение',
 'было',
 'встретить',
 'скептически',
 ',',
 'поскольку',
 'слабый',
 'спутник',
 'оставаться',
 'ненаблюдаемый',
 ',',
 'а',
 'он',
 'масса',
 'должный',
 'быть',
 'быть',
 'достаточно',
 'большой',
 '—',
 'сравнимый',
 'с',
 'мас

In [119]:
# так как spacy выдаёт без пробелов, удаляем их
mystem_lemmas = [i for i in mystem_lemmas if i != ' ']
pymorphy_lemmas = [i for i in pymorphy_lemmas  if i != ' ']

print(len(mystem_lemmas))
print(len(pymorphy_lemmas))
print(len(spacy_lemmas))

299
299
283


In [129]:
for i in range(13):
    if spacy_lemmas[i] != mystem_lemmas[i]:
        print(i, [spacy_lemmas[i], mystem_lemmas[i]])

for j in range(61, len(spacy_lemmas)):
    if spacy_lemmas[j] != mystem_lemmas[j + 3]:
        print(j, [spacy_lemmas[j], mystem_lemmas[j + 3]])

7 ['стать', 'становиться']
9 ['b', 'B']
12 ['b.', 'B']
62 ['прийти', 'приходить']
65 [',', ', ']
70 ['них', 'они']
75 ['.', '. ']
77 ['было', 'быть']
78 ['встретить', 'встречать']
80 [',', ', ']
86 [',', ', ']
88 ['он', 'его']
95 ['—', ' — ']
101 ['проциона', 'процион']
102 [',', ', ']
105 ['\n\n', '\n']
106 ['в', '\n']
107 ['январь', 'в']
108 ['1862', 'январь']
109 ['год', '1862']
110 ['элвин', 'год']
111 ['грэхэм', 'элвин']
112 ['кларк', 'грэхэм']
113 [',', 'кларк']
114 ['юстировать', ', ']
115 ['18-дюймовый', 'юстировать']
116 ['рефрактор', '18']
117 [',', '-']
118 ['самый', 'дюймовый']
119 ['большой', 'рефрактор']
120 ['на', ', ']
121 ['тот', 'самый']
122 ['время', 'большой']
123 ['телескоп', 'на']
124 ['в', 'то']
125 ['мир', 'время']
126 ['(', 'телескоп']
127 ['dearborn', 'в']
128 ['telescope', 'мир']
129 [')', ' (']
130 [',', 'Dearborn']
131 ['впоследствии', 'Telescope']
132 ['поставить', '), ']
133 ['семейный', 'впоследствии']
134 ['фирма', 'поставлять']
135 ['кларк', 'семейный'

In [128]:
for i in range(13):
    if spacy_lemmas[i] != pymorphy_lemmas[i]:
        print(i, [spacy_lemmas[i], pymorphy_lemmas[i]])

for j in range(61, len(spacy_lemmas)):
    if spacy_lemmas[j] != pymorphy_lemmas[j + 3]:
        print(j, [spacy_lemmas[j], pymorphy_lemmas[j + 3]])

0 ['\n', '\r\n']
12 ['b.', 'b']
65 [',', ', ']
70 ['них', 'они']
75 ['.', '. ']
77 ['было', 'быть']
80 [',', ', ']
86 [',', ', ']
95 ['—', ' — ']
101 ['проциона', 'процион']
102 [',', ', ']
105 ['\n\n', '\r\n']
106 ['в', '\r\n']
107 ['январь', 'в']
108 ['1862', 'январь']
109 ['год', '1862']
110 ['элвин', 'год']
111 ['грэхэм', 'элвин']
112 ['кларк', 'грэхэма']
113 [',', 'кларк']
114 ['юстировать', ', ']
115 ['18-дюймовый', 'юстировать']
116 ['рефрактор', '18']
117 [',', '-']
118 ['самый', 'дюймовый']
119 ['большой', 'рефрактор']
120 ['на', ', ']
121 ['тот', 'самый']
122 ['время', 'большой']
123 ['телескоп', 'на']
124 ['в', 'то']
125 ['мир', 'время']
126 ['(', 'телескоп']
127 ['dearborn', 'в']
128 ['telescope', 'мир']
129 [')', ' (']
130 [',', 'dearborn']
131 ['впоследствии', 'telescope']
132 ['поставить', '), ']
133 ['семейный', 'впоследствии']
134 ['фирма', 'поставить']
135 ['кларк', 'семейный']
136 ['в', 'фирма']
137 ['обсерватория', 'кларк']
138 ['чикагского', 'в']
139 ['университет'

Различия есть. Как кажется, spacy в целом чуть хуже справляется

## 5*. LSH (2 балла)

*необязательное задание чтобы получить 10 баллов

Попробуйте искать дубликаты в настоящих текстах. Например, можете взять https://github.com/mannefedov/compling_nlp_hse_course/blob/master/data/anna_karenina.txt или https://github.com/mannefedov/compling_nlp_hse_course/blob/master/data/besy_dostoevsky.txt (или любой другой корпус)

Используйте код из семинара для нахождения кандидатов в дубликаты (шинглы -> минхэш - lsh) и рассчитайте реальную меру Жаккара между полученными кандидатами. Настройте параметры k, num_hash_functions, bands так чтобы результаты получались адекватные (мера Жаккара хотя бы больше нуля). 

(Можете взять 500-1000 текстов если весь корпус обрабатывается слишком долго)

In [161]:
def get_shingles(text, k=10):
    """генерирует список шинглов из строки"""
    shingles = set()
    for i in range(len(text) - k + 1):
        shingle = text[i:i + k]
        shingles.add(shingle)
    return shingles

def hash_string(s):
    """хеширует строку и возвращает число"""
    return int(hashlib.md5(s.encode('utf8')).hexdigest(), 16) 

def generate_hash_functions(k):
    """генерирует k хеш-функций добавляя индекс к строке"""

    functions = []
    for i in range(k):
        functions.append(lambda x, i=i: hash_string(x + str(i)))
    return functions

def compute_minhash_signature(shingles, hash_funcs):
    """вычисляет minhash-сигнатуру для списка шинглов"""
    signature = []
    for hash_func in hash_funcs:
        min_hash = min(hash_func(shingle) for shingle in shingles)
        signature.append(min_hash)
    return signature

def lsh(signatures, bands):
    """Разрезает сигнатуры на куски (bands), и группирует индексы сигнатур по совпадению кусков"""
    buckets = defaultdict(list)
    band_length = len(signatures[0]) // bands
    
    for idx, sig in tqdm(enumerate(signatures)):
        for b in range(0, bands, band_length):
            start = b
            end = start + band_length
            band = tuple(sig[start:end])
            buckets[band].append(idx)
    return buckets

def find_similar_strings(strings_list, k=5, num_hashes=100, bands=20):
    """Finds similar strings using MinHash and LSH."""
    hash_funcs = generate_hash_functions(num_hashes)
    signatures = []
    shingles_list = []

    # каждый текст обрабатывается отдельно
    # находятся шинглы и рассчитываются сигнатуры 
    for string in tqdm(strings_list):
        shingles = get_shingles(string, k)
        shingles_list.append(shingles)
        signature = compute_minhash_signature(shingles, hash_funcs)
        signatures.append(signature)

    # вычисляются кандидаты по кускам сигнатур
    buckets = lsh(signatures, bands)
    candidates = set()
    for bucket in buckets.values():
        if len(bucket) > 1:
            for i in bucket:
                for j in bucket:
                    if i < j:
                        candidates.add((i, j))
    return candidates

def get_jaccard(x, y):
    x_sh = get_shingles(x, 10)
    y_sh = get_shingles(y, 10)

    return len(x_sh & y_sh) / len(x_sh | y_sh)

In [191]:
with open("besy_dostoevsky.txt", encoding="utf-8") as f:
    text = f.read()

sents = sent_tokenize(text)
sents

['Спасибо, что скачали книгу в бесплатной электронной библиотеке Royallib.ru: http://royallib.ru\n\nВсе книги автора: http://royallib.ru/author/dostoevskiy_fedor.html\n\nЭта же книга в других форматах: http://royallib.ru/book/dostoevskiy_fedor/besi.html\n\nПриятного чтения!',
 'Хоть убей, следа не видно,\n\t\tСбились мы, что делать нам?',
 'В поле бес нас водит, видно,\n\t\tДа кружит по сторонам.',
 'Сколько их, куда их гонят,\n\t\tЧто так жалобно поют?',
 'Домового ли хоронят,\n\t\tВедьму ль замуж выдают?',
 'А. Пушкин\n\nТут на горе паслось большое стадо свиней, и они просили Его, чтобы позволил им войти в них.',
 'Он позволил им.',
 'Бесы, вышедши из человека, вошли в свиней; и бросилось стадо с крутизны в озеро и потонуло.',
 'Пастухи, увидя случившееся, побежали и рассказали в городе и по деревням.',
 'И вышли жители смотреть случившееся и, пришедши к Иисусу, нашли человека, из которого вышли бесы, сидящего у ног Иисусовых, одетого и в здравом уме, и ужаснулись.',
 'Видевшие же ра

In [192]:
print(len(sents))
sents = [i for i in sents if len(i) > 10]
print(len(sents))

15291
14691


In [200]:
candidates = find_similar_strings(sents[:5000])

In [203]:
candidates

{(178, 780),
 (514, 1009),
 (523, 1798),
 (523, 1951),
 (698, 3603),
 (698, 4303),
 (698, 4393),
 (698, 4470),
 (754, 870),
 (848, 1085),
 (1133, 3133),
 (1221, 1295),
 (1221, 1306),
 (1233, 1242),
 (1234, 1235),
 (1234, 1236),
 (1235, 1236),
 (1295, 1306),
 (1667, 2579),
 (1739, 1744),
 (1742, 3076),
 (1798, 1951),
 (1834, 4371),
 (1849, 4821),
 (2059, 2160),
 (2148, 2729),
 (2148, 2753),
 (2198, 2232),
 (2219, 2225),
 (2251, 4390),
 (2586, 3376),
 (2621, 3710),
 (2632, 2890),
 (2729, 2753),
 (3452, 3460),
 (3603, 4303),
 (3603, 4393),
 (3603, 4470),
 (3796, 3797),
 (4158, 4203),
 (4303, 4393),
 (4303, 4470),
 (4335, 4384),
 (4335, 4553),
 (4384, 4553),
 (4393, 4470),
 (4539, 4546)}

In [205]:
for i, j in candidates:
    print(i, j)
    print(get_jaccard(sents[i], sents[j]))
    print(sents[i])
    print(sents[j])
    print()

1742 3076
1.0
— Лебядкин?
— Лебядкин?

2251 4390
1.0
— Вы думаете?
— Вы думаете?

1234 1236
0.3333333333333333
Куда же вы?
Куда же вы!

3452 3460
0.8571428571428571
— Сударыня, сударыня!..
— Сударыня, сударыня!

178 780
0.0
Но продолжаю.
— продолжал смеяться Nicolas.

4335 4384
1.0
— спросил он.
— спросил он.

2586 3376
0.3225806451612903
Маврикий Николаевич!
Маврикий Николаевич поклонился и вышел.

2219 2225
0.5434782608695652
Кто победит боль и страх, тот сам бог будет.
Кто победит боль и страх, тот сам станет бог.

1234 1235
1.0
Куда же вы?
Куда же вы?

2148 2729
1.0
— засмеялся я.
— засмеялся я.

2059 2160
0.4
Я покраснел.
Он покраснел.

1798 1951
0.25806451612903225
— Помилуйте, Степан Трофимович!
Степан Трофимович, это вы?

698 4303
0.3611111111111111
Николай Всеволодович усмехнулся.
Николай Всеволодович упорно молчал.

3603 4303
0.3793103448275862
— Николай Всеволодович!
Николай Всеволодович упорно молчал.

4539 4546
0.02666666666666667
Отперев дверцу, он подал ключ Николаю Всев

В основоном Жаккардово расстояние получается >0, а там, где 0 (или почти 0, как в 4539 4546), предложения действительно не схожие, т.е. дедупликация удалась, согласно условию